beacon_node/genesis/src/eth1_genesis_service.rs

```diff
@@ -0,0 +1,374 @@
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Produces an eth2 genesis `BeaconState` from the given `eth1_block`.
+    ///
+    /// ## Returns
+    ///
+    /// - Ok(genesis_state) if all went well.
+    /// - Err(e) if the given `eth1_block` was not a viable block to trigger genesis or there was
+    /// an internal error.
+    fn genesis_from_eth1_block<E: EthSpec>(
+        &self,
+        eth1_block: Eth1Block,
+        spec: &ChainSpec,
+    ) -> Result<BeaconState<E>, String> {
+        let deposit_logs = self
+            .core
+            .deposits()
+            .read()
+            .cache
+            .iter()
+            .take_while(|log| log.block_number <= eth1_block.number)
+            .map(|log| log.deposit_data.clone())
+            .collect::<Vec<_>>();
+
+        let genesis_state = initialize_beacon_state_from_eth1(
+            eth1_block.hash,
+            eth1_block.timestamp,
+            genesis_deposits(deposit_logs, &spec),
+            &spec,
+        )
+        .map_err(|e| format!("Unable to initialize genesis state: {:?}", e))?;
+
+        if is_valid_genesis_state(&genesis_state, &spec) {
+            Ok(genesis_state)
+        } else {
+            Err("Generated state was not valid.".to_string())
+        }
+    }
+
+    /// A cheap (compared to using `initialize_beacon_state_from_eth1) method for determining if some
+    /// `target_block` will trigger genesis.
+    fn is_valid_genesis_eth1_block<E: EthSpec>(
+        &self,
+        target_block: &Eth1Block,
+        spec: &ChainSpec,
+    ) -> Result<bool, String> {
+        if target_block.timestamp < spec.min_genesis_time {
+            Ok(false)
+        } else {
+            let mut local_state: BeaconState<E> = BeaconState::new(
+                0,
+                Eth1Data {
+                    block_hash: Hash256::zero(),
+                    deposit_root: Hash256::zero(),
+                    deposit_count: 0,
+                },
+                &spec,
+            );
+
+            local_state.genesis_time = target_block.timestamp;
+
+            self.deposit_logs_at_block(target_block.number)
+                .iter()
+                // TODO: add the signature field back.
+                //.filter(|deposit_log| deposit_log.signature_is_valid)
+                .map(|deposit_log| Deposit {
+                    proof: vec![Hash256::zero(); spec.deposit_contract_tree_depth as usize].into(),
+                    data: deposit_log.deposit_data.clone(),
+                })
+                .try_for_each(|deposit| {
+                    // No need to verify proofs in order to test if some block will trigger genesis.
+                    const PROOF_VERIFICATION: bool = false;
+
+                    process_deposit(
+                        &mut local_state,
+                        &deposit,
+                        spec,
+                        PROOF_VERIFICATION,
+                        // TODO: disable signature verification
```

Comment:
This could be made much faster by pre-verifying block signatures, filtering out those with bad signatures, then disabling signature verification in this function.